# Mini-Demo's

## Simple Linear Regression

A model which uses linear weights to optimize a model that fits the
given data. The number of tunable parameters is equal to the number
of variables plus 1.

<u>Linear Regression Models</u>
* Simple Linear Regression
* Ridge Regression
* Lasso Regression
* Elastic Net Regression
* Random Forest Regression

In [1]:
import numpy as np
from dask.dataframe.dispatch import tolist
from sklearn.linear_model import LinearRegression

X = np.array([1, 5, 3, 10])
y = np.array([3, 8, 11, 11])

reg = LinearRegression()

In [2]:
print("X\t\t:", X)
print("X.shape\t\t:", X.shape)

print("\n---------------\n")

print("X.argmax()\t:", X.argmax())
print("X.max()\t\t:", X.max())

X.argsort()


X		: [ 1  5  3 10]
X.shape		: (4,)

---------------

X.argmax()	: 3
X.max()		: 10


array([0, 2, 1, 3])

In [4]:
reg.fit(X, y)

ValueError: Expected 2D array, got 1D array instead:
array=[ 1  5  3 10].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.


-----------------------

Say we have the following:

In [12]:
import pandas as pd

#t1 = pd.DataFrame([66,54,69,73,61,62],columns=['A1'])
t1 = pd.DataFrame({"A1":[66,     54,   69,   73,   61,   62],
                   "A2":[33.6, 26.6, 23.3, 28.1, 43.1, 25.6]})

t1

1st) To calculate the covariance matrix, make table **zero mean**.

2nd) Find mean of all the data & subtract it off

<br>

Recall:

- Sample Mean: &nbsp;&nbsp; $\bar{x} = \frac{\sum x_{i}}{n}$
- Sample Variance: &nbsp;&nbsp; $s^{2} = \frac {\sum(x_{i}-\bar{x})^2}{n-1}$

In [17]:
a1_mean = t1["A1"].mean()
a2_mean = t1["A2"].mean()

print("A1 Mean:", a1_mean)
print("A2 Mean:", a2_mean)

A1 Mean: 64.16666666666667
A2 Mean: 30.049999999999997


We now have the following:

$\bar{x}_{A1} = 64.166667 = \frac{\sum A1_{i}}{6} = \frac{66 + 54 + 69 + 73 + 61 + 62}{6}$

$\bar{x}_{A2} = 30.049997 = \frac{\sum A2_{i}}{6} = \frac{33.6 + 26.6 + 23.3 + 28.1 + 43.1 + 25.6}{6}$

<br>

From here we can then calculate the Sample Variance

$s^{2} = \frac {\sum(A1_{i} - 64.1666667)^2}{n}$

$s^{2} = \frac {\sum(A2_{i} - 30.0499997)^2}{n}$


In [54]:
# Now we subtract the mean from each value
t2 = pd.DataFrame({"A1": [x - a1_mean for x in t1["A1"]],
                   "A2": [x - a2_mean for x in t1["A2"]]})

t2

A1     A2
0   1.833333   3.55
1 -10.166667  -3.45
2   4.833333  -6.75
3   8.833333  -1.95
4  -3.166667  13.05
5  -2.166667  -4.45

In [70]:
t2["A1"].sum()

t3 = pd.DataFrame({"A1": [pow(a1,2) for a1 in t2["A1"]],
                   "A2": [pow(a2,2) for a2 in t2["A2"]]})

t3

A1        A2
0    3.361111   12.6025
1  103.361111   11.9025
2   23.361111   45.5625
3   78.027778    3.8025
4   10.027778  170.3025
5    4.694444   19.8025

In [78]:
a_1 = (t3["A1"].sum()) / 6
a_2 = (t3["A2"].sum()) / 6

print("A1:\t", a_1)
print("A2:\t", a_2)

(a_1 + a_2) / 5

A1:	 37.13888888888889
A2:	 43.99583333333334


16.226944444444445


---------------------------------------------




* Matrix Multiplication
* Dot Product
* Transposition

In order to do dot product, 


In [ ]:
import numpy as np
import pandas as pd

